### PR TITLE
Refactor: Move streaming supervisor methods to `SeekableStreamSupervisor`

### PR DESCRIPTION
**Pull request by abhishekrb19**
The current `Supervisor` interface is primarily focused on streaming use cases. However, as we introduce supervisors for non-streaming use cases, such as the recently added `CompactionSupervisor` (and the upcoming `BatchSupervisor`), certain operations like resetting offsets, checkpointing, task group handoff, etc., are not really applicable to non-streaming use cases. 

So the methods are split between:
- `Supervisor`: common methods that are applicable to both streaming and non-streaming use cases
- `SeekableStreamSupervisor`: `Supervisor` + streaming-only operations. The existing streaming-only overrides exist along with the new abstract method `public abstract LagStats computeLagStats()`, for which custom implementations already exist in the concrete types

This PR is primarily a refactoring change with minimal functional adjustments (e.g., throwing an exception in a few places in `SupervisorManager` when the supervisor isn't the expected `SeekableStreamSupervisor` type).

This PR has:

- [x] been self-reviewed.
- [x] added documentation for new or modified features or behaviors.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.

### CODE REVIEW

File: indexing-service/src/main/java/org/apache/druid/indexing/overlord/supervisor/SupervisorManager.java [hunk 1]
```diff
@@ -142,7 +143,15 @@ public boolean handoffTaskGroupsEarly(String id, List<Integer> taskGroupIds)
     if (supervisor == null || supervisor.lhs == null) {
       return false;
     }
-    supervisor.lhs.handoffTaskGroupsEarly(taskGroupIds);
+    if (!(supervisor.lhs instanceof StreamSupervisor)) {
+      throw new UOE(
```
Comment (kfaraz) [inline indexing-service/src/main/java/org/apache/druid/indexing/overlord/supervisor/SupervisorManager.java:147]: Throw DruidException instead?

File: indexing-service/src/main/java/org/apache/druid/indexing/overlord/supervisor/SupervisorManager.java [hunk 2]
```diff
@@ -277,10 +286,20 @@ public boolean resetSupervisor(String id, @Nullable DataSourceMetadata resetData
...
 
+    if (!(supervisor.lhs instanceof StreamSupervisor)) {
+      throw new UOE(
+          "Reset operation is not supported by the supervisor[%s] of type[%s].",
+          id,
+          supervisor.rhs.getType()
+      );
+    }
+
+    final StreamSupervisor streamSupervisor = (StreamSupervisor) supervisor.lhs;
```
Comment (kfaraz) [inline indexing-service/src/main/java/org/apache/druid/indexing/overlord/supervisor/SupervisorManager.java:297]: This logic can be captured in a method and reused, something like:

```
private SeekableStreamSupervisor requireStreamSupervisor(String supervisorId, String operation)
{
    Pair supervisor = supervisors.get(id);
    if (supervisor instanceof SeekableStreamSupervisor) {
        // cast and return
    } else {
       // throw DruidException with right message about operation
    }
}
```

**Review by kfaraz — COMMENTED** [review]
Approach looks, left minor suggestions.

Also, maybe we don't need `StreamSupervisor`, we may just move the relevant methods to `SeekableStreamSupervisor`.

File: indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/SeekableStreamSupervisor.java [hunk 3]
```diff
@@ -4170,7 +4167,7 @@ public SeekableStreamSupervisorIOConfig getIoConfig()
     return ioConfig;
   }
 
-  @Override
+
   public void checkpoint(int taskGroupId, DataSourceMetadata checkpointMetadata)
```
Comment (kfaraz) [inline indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/SeekableStreamSupervisor.java:4171]: Please add the javadocs originally present in `Supervisor` interface.

**Review by kfaraz — APPROVED** [review]
Requires a couple of javadocs, rest looks good. 👍🏻

File: indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/SeekableStreamSupervisor.java (hunk 3 above)
Comment (abhishekrb19) [reply indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/SeekableStreamSupervisor.java:4184]: Done, moved the javadocs too.

**Pull request merged**